### DISCUSSION TITLE
Define an event as the sum of two events

### DISCUSSION THREAD

**Discussion by jbpfran** [RailsEventStore/rails_event_store#1615] [Q&A] [answered]
Hello,

I was wondering how to handle the following scenario :
- a user change its profile's avatar, raising a ProfileUpdated event
- the new avatar is processed in a background job to create derivatives (handled by Shrine), raising a AvatarUpdated event
- you need to propagate these changes only when both event are there

So is there a way to raise a event and wait for another event to launch an action when both are there ?

Thanks for your help

**Comment by mostlyobvious** [comment, marked as answer]
Hello,

what you've described fits perfectly into a definition of a Process. There are certain condition to be met over time (expressed with `ProfileUpdated` and `AvatarUpdated` events). When they all happen, a barrier breaks and we propagate changes.

A component that handles such processes is called a [Process Manager](https://blog.arkency.com/tags/process-manager/). It takes events as input (something happened), issuing a command when all conditions are met (do something):

```
f(events) -> command
```

In RailsEventStore this would be an event handler. Such handler needs to be subscribed to both events you've mentioned. Also both events need to have something in common, i.e. an identifier of entity they describe (like `profile_id`), so that we could associate them to the particular instance of the process. 

Another thing to implement in this handler is state management, as we need to keep track of the process over time. State can be event-sourced — for example by [linking events related to a process into a given stream](https://blog.arkency.com/process-managers-revisited/#linking_events_to_the_rescue) and checking whether state is "complete" each time new event arrives. Alternatively state can be tracked differently, i.e. [in a database table](https://blog.arkency.com/2017/06/dogfooding-process-manager/#classical_example).

When testing, it is useful to [check all permutations](https://blog.arkency.com/2016/06/cover-all-test-cases-with-permutation/) of input events. They may arrive out of order due to messaging infrastructure. Yet that should make no difference to the process.
